internal/test/builder/builders.go

```diff
@@ -602,6 +787,82 @@ func (i *TestInfrastructureMachineTemplateBuilder) Build() *unstructured.Unstruc
+	obj := &unstructured.Unstructured{}
+	obj.SetName(name)
+	obj.SetNamespace(namespace)
+	obj.SetAPIVersion(InfrastructureGroupVersion.String())
+	obj.SetKind(GenericInfrastructureMachineTemplateKind)
+	// Set the mandatory spec fields for the object.
+	setSpecFields(obj, map[string]interface{}{"spec.template.spec": map[string]interface{}{}})
+	return &InfrastructureMachinePoolTemplateBuilder{
+		obj,
+	}
+}
+
+// WithSpecFields sets a map of spec fields on the unstructured object. The keys in the map represent the path and the value corresponds
+// to the value of the spec field.
+//
+// Note: all the paths should start with "spec."
+//
+//	Example map: map[string]interface{}{
+//	    "spec.version": "v1.2.3",
+//	}.
+func (i *InfrastructureMachinePoolTemplateBuilder) WithSpecFields(fields map[string]interface{}) *InfrastructureMachinePoolTemplateBuilder {
+	setSpecFields(i.obj, fields)
+	return i
+}
+
+// Build takes the objects and variables in the  InfrastructureMachineTemplateBuilder and generates an unstructured object.
+func (i *InfrastructureMachinePoolTemplateBuilder) Build() *unstructured.Unstructured {
+	return i.obj
+}
+
+// TestInfrastructureMachinePoolTemplateBuilder holds the variables and objects needed to build an TestInfrastructureMachinePoolTemplate.
+type TestInfrastructureMachinePoolTemplateBuilder struct {
+	obj *unstructured.Unstructured
+}
+
+// TestInfrastructureMachinePoolTemplate creates an TestInfrastructureMachinePoolTemplateBuilder with the given name and namespace.
+func TestInfrastructureMachinePoolTemplate(namespace, name string) *TestInfrastructureMachinePoolTemplateBuilder {
+	obj := &unstructured.Unstructured{}
+	obj.SetName(name)
+	obj.SetNamespace(namespace)
+	obj.SetAPIVersion(InfrastructureGroupVersion.String())
+	obj.SetKind(TestInfrastructureMachineTemplateKind)
+	// Set the mandatory spec fields for the object.
+	if err := unstructured.SetNestedField(obj.Object, map[string]interface{}{}, "spec", "template", "spec"); err != nil {
+		panic(err)
+	}
+	return &TestInfrastructureMachinePoolTemplateBuilder{
+		obj,
+	}
+}
+
+// WithSpecFields sets a map of spec fields on the unstructured object. The keys in the map represent the path and the value corresponds
+// to the value of the spec field.
+//
+// Note: all the paths should start with "spec."; the path should correspond to a field defined in the CRD.
+//
+//	Example map: map[string]interface{}{
+//	    "spec.version": "v1.2.3",
+//	}.
+func (i *TestInfrastructureMachinePoolTemplateBuilder) WithSpecFields(fields map[string]interface{}) *TestInfrastructureMachinePoolTemplateBuilder {
+	setSpecFields(i.obj, fields)
+	return i
+}
+
+// Build takes the objects and variables in the  InfrastructureMachineTemplateBuilder and generates an unstructured object.
+func (i *TestInfrastructureMachinePoolTemplateBuilder) Build() *unstructured.Unstructured {
+	return i.obj
+}
+
```

Comment:
```suggestion
```

I think we can drop this whole part as it is the same as https://github.com/kubernetes-sigs/cluster-api/pull/9346/files#diff-30b8ed7bcbffa0687acaccab71710d58dab625a59a40a38888069eb870b41c17R714-R789 except the builder name. The resulting unstructured is of same type.

Reply:
Apologies if I'm missing something. Are you suggesting that folks should use the TestInfraMachineTemplateBuilder when they need a TestInfraMachinePoolTemplate?

Reply:
I see your point that it's just the same but I think it would be problematic semantically in the unit test to create infrastructuremachinetemplates and use them as inframachinepooltemplates.

Reply:
It is at the end a `GenericInfrastructureMachineTemplateKind` in verison `InfrastructureGroupVersion.String()`.
```go
	obj.SetAPIVersion(InfrastructureGroupVersion.String())
	obj.SetKind(GenericInfrastructureMachineTemplateKind)
```

Shouldn't the builder reflect the type?

Reply:
But I'm also okay with duplicating if that is the real intent.

Reply:
otherwise the new struct maybe should have

```go
	obj.SetAPIVersion(InfrastructureGroupVersion.String())
	obj.SetKind(GenericInfrastructureMachinePoolTemplateKind)
```

Reply:
Ah got it. The kind should be TestInfrastructureMachinePoolTemplateKind (not generic)

Reply:
I intitially also went ahead and simply used a infrastructuremachinetemplate as a inframachinepooltemplate in my unit tests. But to be semantically safer I would suggest to:
- Duplicate the part (like it currently is in this PR)
- Reintroduce GenericInfrastructureMachinePoolTemplateKind and change it to `obj.SetKind(GenericInfrastructureMachinePoolTemplateKind)` as @chrischdi suggested

Reply:
So tl;dr we need a separate builder for InfrastructureMachinePoolTemplate not a duplicate one to what we already have :)